File: Jacobi.m
function x = Jacobi(A)
  ## 运用向量化思路实现雅可比迭代，已设定参数默认值，可直接键入 Jacobi 运行
  ## A：增广矩阵

  ## 设定参数默认值
  if ~exist('A')
    A = [-4 1 1 1 1; 1 -4 1 1 1; 1 1 -4 1 1; 1 1 1 -4 1];
  end

  [M N] = size(A);

  ## 对于每一个x_i，同时计算，故提取a_ii和a_ij
  aii = diag(A(1:end, 1:end-1)); # 提取对角线元素
  aij = A(1:end, 1:end-1) - eye(M) .* aii; # 满足i≠j，即去除对角线元素

  ## 初始化
  x0 = zeros(M, 1);
  xk = ones(M, 1);

  ## 开始迭代
  while max(abs(x0 - xk)) >= 0.00001

    xk = x0;
  # disp(xk);
    x0 = (1./aii) .* (A(:, end) - aij * xk) ;
  # disp(x0);

  end
  x = x0;

end

File: Jacobi.py
import numpy as np

A = None
def Jacobi(A=[[-4, 1, 1, 1, 1],
             [1, -4, 1, 1, 1],
             [1, 1, -4, 1, 1],
             [1, 1, 1, -4, 1]]):
    """ 雅可比迭代向量化
        返回值：方程组解
    Args:
        A (list): 增广矩阵
    """ 
    A = np.array(A, dtype=float)
    [M, N] = np.shape(A)
    
    aii = np.diag(A[:,:-1]).reshape(-1, 1) # reshape转换成列向量
    aij = A[:,:-1] - np.eye(M) * aii # 始终注意python索引从0开始。左闭右开
    
    x0 = np.zeros((M, 1)) # 双重(())
    xk = np.ones((M, 1))
    
    while np.max(np.abs(x0 - xk)) >= 0.00000001: # np.max 和 np.abs类型都为narray
        
        xk = x0
        x0 = (np.divide(1, aii)) * (A[:, -1].reshape(-1, 1) - np.dot(aij, xk))
        
    return x0

if __name__ == '__main__':

    A = [[20, 2, 3, 24],[1, 8, 1, 12],[2, -3, 15, 30]]
    print(Jacobi(A))